### ISSUE TITLE
access token expiry time

### ISSUE THREAD

**Issue by gelmanovic** [snobu/destreamer#96]
How come the access is only granted for less than an hour? On OBS software I don't need to ever login again but just once in the beginning (even after closing the application).
This wouldn't be much of a problem considered the many upsides if it wasn't for the fact that I get no notification nor on the moment in which I should be prompted to login again nor about what link from the URLs txt file it stopped at nor do I get to keep seeing the git command line interface when I'm back on my pc every once in a while if the access token has already expired.
Thank you all.

**Comment by kylon** [comment]
because the refresh is missing right now, please use the old version if you are planning to download lots of videos with no interaction

**Comment by gelmanovic** [comment]
is there just one other older version?

**Comment by kylon** [comment]
https://github.com/snobu/destreamer/tree/dev

the dev branch

**Comment by gelmanovic** [comment]
thanks man, mucho apreciated de verdad

**Comment by gelmanovic** [comment]
can you please explain where should be located the path you mentioned?
"youtube-dl: https://ytdl-org.github.io/youtube-dl/download.html, you'll need a fairly recent version that understands encrypted HLS streams. This needs to be in your $PATH"

**Comment by snobu** [comment]
Yoour environment's $PATH or place it in the same directory with destreamer.js

**Comment by gelmanovic** [comment]
I can't seem to clone this branch
am I doing something wrong?
git clone https://github.com/snobu/destreamer/tree/dev

**Comment by snobu** [comment]
`git clone -b dev https://github.com/snobu/destreamer`

**Comment by gelmanovic** [comment]
ok cool, thanks bye

**Comment by snobu** [comment]
We've just merged a fix for the token refresh into master. You may want to give that another go, please let us know if you're still seeing 401 Unauthorized exceptions if you do.

**Comment by gelmanovic** [comment]
Oh that's so nice of you. I am already running the dev branch with no hassle, better yet I find it even faster at downloading.
Anyway I shall be checking that one out too. But are you saying that also in the master branch there won't be no need to login every now and then?

**Comment by snobu** [comment]
You'll need to login once every hour if destreamer isn't running. If it's running (downloading a list of videos for example) it will now refresh tokens for you so no interaction is necessary.

**Comment by gelmanovic** [comment]
Definetly loved this last edit of yours LMFAOL

**Comment by gelmanovic** [comment]
gotten the 401, now what

**Comment by gelmanovic** [comment]
I noticed files sizes were remarkably larger with the dev branch

**Comment by gelmanovic** [comment]
Oh and I don't get what you mean by running it in windows cmd... when I do it still calls git bash where it prints on screen "sdtout is not a tty"

**Comment by kylon** [comment]
> gotten the 401, now what

can you give us more info

> I noticed files sizes were remarkably larger with the dev branch, only a few times they were exactly the same

yeah, bad compression / no compression


> Oh and I don't get what you mean by running it in windows cmd... when I do it still calls git bash where it prints on screen "sdtout is not a tty"

try powershell, your path is probably messed

**Comment by gelmanovic** [comment]
> can you give us more info
afraid not, by the way you spoke last night I though you already knew all there were to know about the error

> yeah, bad compression / no compression
now it's killing the process 
![timeouterror](https://user-images.githubusercontent.com/57871063/80351869-51032a80-8873-11ea-9f5a-228c6bb84674.png)

**Comment by gelmanovic** [comment]
eventually made it to login but now it's stuck
![stuck](https://user-images.githubusercontent.com/57871063/80352766-adb31500-8874-11ea-99f3-d27233243547.png)

**Comment by gelmanovic** [comment]
Actually I know I shouldn't plea for aid on deprecated one but I also accepted to give the latest one a go with no success.

**Comment by snobu** [comment]
For the master branch, could you please confirm you have this commit in your clone of the repo?

```
Add mid-download token refresh, fix headers length overflow bug (#93)
Commit id 042e79d57f4fb31cc3bf326a89db37160b6e1941
```

Use `git log`.

**Comment by lukaarma** [comment]
It's your connection good/stable?
Because before you had a problem with a page load timeout of 30s and now ffmpeg hangs waiting on a response (nothing we can go about I'm sorry to tell you)

What problem did you have with the new master branch that the old Dev one solves?

**Comment by gelmanovic** [comment]
> For the master branch, could you please confirm you have this commit in your clone of the repo?
> 
> ```
> Add mid-download token refresh, fix headers length overflow bug (#93)
> Commit id 042e79d57f4fb31cc3bf326a89db37160b6e1941
> ```
> 
> Use `git log`.

I have that


> It's your connection good/stable?
> Because before you had a problem with a page load timeout of 30s and now ffmpeg hangs waiting on a response (nothing we can go about I'm sorry to tell you)
> 
> What problem did you have with the new master branch that the old Dev one solves?

I usually don't experience any troubles with connectivity.

**Comment by gelmanovic** [comment]
It basically went back to previous functioning mode, with git bash disappearing when token expired and no alert or prompt to login again

**Comment by snobu** [comment]
If you still see a 401 Unauthorized exception being thrown and then destreamer exiting, try changing the refresh interval to 600 (seconds) and rebuild.
https://github.com/snobu/destreamer/blob/master/src/destreamer.ts#L28

You should also launch via destreamer.cmd in cmd.exe, not the .sh.

Your tenant may issue tokens with shorter than one hour validity.

**Comment by gelmanovic** [comment]
I'm sorry I couldn't find on google how to rebuild, please advice.. Sorry for being such a pain in the butt

**Comment by snobu** [comment]
Just follow the readme -
https://github.com/snobu/destreamer#how-to-build
```
npm run build
```

**Comment by gelmanovic** [comment]
done that yet still the same 401

**Comment by snobu** [comment]
Ah yes, my bad, i ran my end to end test again and got the 401 as well. I'll get a proper fix this time around.

Also, you know... :)
<img src="https://webdevstudios.com/wp-content/uploads/2015/03/its-free-software.gif" />

**Comment by gelmanovic** [comment]
uhmm..
I was about to buy some software for doing this, if only they could deal with MS stream which none of them actually happen to do.
I'd be eager to pay you some wage if I could just sit back the whole time, inputting the commands once a day.

**Comment by gelmanovic** [comment]
Didn't want to disturb you but I'm seeing last unclosed issue was created in september. So maybe it's not closed just because user disappeared, leading me to believe otherwise you would normally get them to be closed fast enough.
I can wait anytime (well in fact I wouldnt go past june) but just wanted to make sure you don't forget me.

**Comment by snobu** [comment]
Try to run with `-x` (no experiements), i have a hunch the 401 Unauthorized may come from the thumbnail rendering rather than download logic.

**Comment by gelmanovic** [comment]
I always wrote the -x specific when launching.
It is failing right after the third video, not after the 100th.

**Comment by gelmanovic** [comment]
Well at times I get to download even 5-6 videos in a row but I'm assuming you wouldn't have included the URLs file option if someone were to only put 5-6 URLs in there.

**Comment by ahmedhamdy90** [comment]
Is there a solution to this issue yet? As the login expires every hour, so if I am in a session and want to download part 2 of it I need to re-login.

It is not a big deal, but I think it is a nice to have feature to increase the access token alive time by default or using a parameter.

**Comment by elbarto404** [comment]
the program is awsome!🥇  but this expiration token problem makes the master version unusefull in some cases... please fix it 🙏

**Comment by J-Savage** [comment]
It times out for me after an hour as well, but if you just add `--skip true` it skips the ones that are completed and carries on where it left off. As @snobu said above, its intended for downloading a handful of videos, but this wont work if you are trying to download thousands of videos in one list.

**Comment by gelmanovic** [comment]
> It times out for me after an hour as well, but if you just add `--skip true` it skips the ones that are completed and carries on where it left off. As @snobu said above, its intended for downloading a handful of videos, so this wont work if you are trying to download thousands of videos in one list.

So where did you find such a command?
Anyway, as I said above...

> It is failing right after the third video, not after the 100th.

So the problem is still there

**Comment by J-Savage** [comment]
The command was on the readme file, under Usage. And yes, for this to work you need to leave the completed videos in the target folder. It shouldn't stop after any amount of downloads, just however many downloads you can get before it times out.

**Comment by snobu** [comment]
Bad news chaps, the API endpoint aptly named `/refreshtoken`... well.. doesn't actually refresh anything. I believe it's a deprecated thing and they fully rely on the IdP to do the session keep alive (something we can't do without a browser currently).

I kept blaming myself there's a bug in the client code calling that API but took it outside destreamer and used basic REST API calls and the cookies i got back did not have their lifetime extended at all.

We'll probably need to keep a puppeteer instance running (hidden maybe?) so it can handle the _refreshments_.

I'm still at the drawing board.

Meanwhile use the `dev` branch if you really need batch downloads.

**Comment by snobu** [comment]
Alright, need a bit of help here.
Can any of you please open an Incognito browser window and go to [this URL][1]?

Log in, and let me know if you either get a TIMEOUT error mentioning something about a `http://localhost:4000` or you get a boring `Your admin needs to grant consent for this app blah blah`.

[1]: https://login.microsoftonline.com/common/oauth2/authorize?client_id=e63af00a-db93-472a-a4ea-1e6f4321a934&response_mode=form_post&response_type=code+id_token&scope=openid+profile&state=OpenIdConnect.AuthenticationProperties%3dcGxCvc7XOC6-U1_OiQiQqISCrnb6nMdq98owGrLdG1RpQFU3rq2w2B0zl29IUOnr9Vgl7qXsTYfVGOy-gkCm3l99_Rus1nX81SIyML5T71lenFNg1oRHVUeerIP9BM1Fs1QicSrkX3nuOzWGxKeRq9h9PRcXlBZ7zPnSuZWQI-DYN49LYZ2oMis-dJMw1l2vNBjZCx3Tk_cRlrHj1z8Dxnq8FubkY_UAsOOb63PIxilSo2xL&nonce=637246578251270009.ZTE0NTMxMWItY2NhZS00NjY2LWJiNWYtMmMwNmVjYjc5NWYxMmQ2N2M2YzEtMTYxOS00MjY0LTgzNmItZjhmN2U1OTllZTAw&nonceKey=OpenIdConnect.nonce.mu79tKAY8Z9fWdBscxjlgbxa%2bldAzE0czmpj0VDgUgs%3d&site_id=500453&redirect_uri=http%3a%2f%2flocalhost:4000%2f&post_logout_redirect_uri=https%3a%2f%2fproducts.office.com%2fmicrosoft-stream&msafed=0

**Comment by gelmanovic** [comment]
"Safari can't load the page because connection to server has failed"

**Comment by gelmanovic** [comment]
Or do you want a console log?

**Comment by snobu** [comment]
That's actually good news. Can you use a browser that's *not* Safari please so we can see the full error message?

**Comment by gelmanovic** [comment]
![2](https://user-images.githubusercontent.com/57871063/81495859-adbf0600-92b3-11ea-9838-3802b4c9762e.png)
![1](https://user-images.githubusercontent.com/57871063/81495860-ae579c80-92b3-11ea-8d12-959f2dc509b2.png)

**Comment by snobu** [comment]
Excellent, can you now please go to the Network tab and see if the request to `localhost` has a base64 encoded token in the body?

**Comment by gelmanovic** [comment]
[destr.txt](https://github.com/snobu/destreamer/files/4606354/destr.txt)

**Comment by gelmanovic** [comment]
I guess so

**Comment by snobu** [comment]
Appreciated!

**Comment by gelmanovic** [comment]
So i see this is overcome now with latest build, is that correct?
Will it suffice to just unbuild and clone from scratch? I am not into patching much.

**Comment by snobu** [comment]
Unfortunately, this is still an outstanding bug. Appreciate the help here, but i'm afraid doing OAuth is not going to work for everyone so we need to go back to the drawing board.

**Comment by gelmanovic** [comment]
So I lost myself along the way...
Where are we? There've been a bunch of updates. 
Was the unmatched URLs-filepath order fixed?

**Comment by gelmanovic** [comment]
right now it doesn't crash when token expires any more, better still it doesn't even prompt me for logging in again
